File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;
    static int N, M, K;
    static long[] arr;
    static long[] tree;

    static void build(int node, int start, int end) {
        if (start == end) {
            tree[node] = arr[start];
            return;
        }

        int mid = (start + end) / 2;

        build(node * 2, start, mid);
        build(node * 2 + 1, mid + 1, end);
        tree[node] = tree[node * 2] + tree[node * 2 + 1];
    }

    static void update(int idx, long val, int node, int start, int end) {
        if (idx < start || end < idx)
            return;

        if (start == end) {
            tree[node] = val;
            return;
        }

        int mid = (start + end) / 2;

        update(idx, val, node * 2, start, mid);
        update(idx, val, node * 2 + 1, mid + 1, end);
        tree[node] = tree[node * 2] + tree[node * 2 + 1];
    }

    static long query(int left, int right, int node, int start, int end) {
        if (right < start || end < left)
            return 0;

        if (left <= start && end <= right) {
            return tree[node];
        }

        int mid = (start + end) / 2;

        return query(left, right, node * 2, start, mid) + query(left, right, node * 2 + 1, mid + 1, end);
    }

    static public void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(st.nextToken());
        arr = new long[N];
        tree = new long[4 * N];
        for (int i = 0; i < N; i++) {
            arr[i] = Long.parseLong(br.readLine());
        }
        build(1, 0, N - 1);

        for (int i = 0; i < M + K; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            if (a == 1) {
                int b = Integer.parseInt(st.nextToken());
                long c = Long.parseLong(st.nextToken());
                update(b - 1, c, 1, 0, N - 1);
            } else {
                int b = Integer.parseInt(st.nextToken());
                int c = Integer.parseInt(st.nextToken());
                bw.write(query(b - 1, c - 1, 1, 0, N - 1) + "\n");
            }
        }
        bw.close();
    }
}

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.py
import sys

input = lambda: sys.stdin.readline().rstrip()


def build(node, start, end):
    if start == end:
        tree[node] = arr[start]
        return

    mid = (start + end) // 2

    build(node * 2 + 1, start, mid)
    build(node * 2 + 2, mid + 1, end)
    tree[node] = tree[node * 2 + 1] + tree[node * 2 + 2]


def update(idx, val, node, start, end):
    if idx < start or end < idx:
        return
    if start == end:
        tree[node] = val
    else:
        mid = (start + end) // 2
        update(idx, val, node * 2 + 1, start, mid)
        update(idx, val, node * 2 + 2, mid + 1, end)
        tree[node] = tree[node * 2 + 1] + tree[node * 2 + 2]


def query(left, right, node, start, end):
    if right < start or end < left:
        return 0

    if left <= start and end <= right:
        return tree[node]

    mid = (start + end) // 2

    return query(left, right, node * 2 + 1, start, mid) + query(left, right, node * 2 + 2, mid + 1, end)


if __name__ == '__main__':
    N, M, K = map(int, input().split())
    arr = [int(input()) for _ in range(N)]
    tree = [0] * (N * 4)
    build(0, 0, N - 1)

    for _ in range(M + K):
        a, b, c = map(int, input().split())
        if a == 1:
            update(b - 1, c, 0, 0, N - 1)
        else:
            print(query(b - 1, c - 1, 0, 0, N - 1))